backend/migrations/2023-07-03-165000_heatmap/up.sql

```diff
@@ -0,0 +1,241 @@
+DECLARE
+    score SCORE;
+    map_geometry GEOMETRY(POLYGON, 4326);
+    point GEOMETRY;
+    bbox GEOMETRY;
+    num_cols INTEGER;
+    num_rows INTEGER;
+    x_pos INTEGER;
+    y_pos INTEGER;
+    plant_relation RECORD;
+BEGIN
+    -- Makes sure the layers exists and fits to the map
+    FOR i IN 1..array_length(p_layer_ids, 1) LOOP
+        IF NOT EXISTS (SELECT 1 FROM layers WHERE id = p_layer_ids[i] AND map_id = p_map_id) THEN
+            RAISE EXCEPTION 'Layer with id % not found on map', p_layer_ids[i];
+        END IF;
+    END LOOP;
+    -- Makes sure the plant exists
+    IF NOT EXISTS (SELECT 1 FROM plants WHERE id = p_plant_id) THEN
+        RAISE EXCEPTION 'Plant with id % not found', p_plant_id;
+    END IF;
+
+    -- INTO STRICT makes sure the map exists. Does have to be explicitly checked as bounding box calculation would error anyways.
+    SELECT geometry FROM maps WHERE id = p_map_id INTO STRICT map_geometry;
+
+    -- Calculate the number of rows and columns based on the map's size and granularity
+    num_cols := FLOOR((x_max - x_min) / granularity); -- Adjusted for granularity
+    num_rows := FLOOR((y_max - y_min) / granularity); -- Adjusted for granularity
+
+    -- Calculate the score for each point on the heatmap
+    FOR i IN 0..num_cols-1 LOOP
+        FOR j IN 0..num_rows-1 LOOP
+            -- i and j do not represent coordinates. We need to adjust them to actual coordinates.
+            x_pos := x_min + (i * granularity) + (granularity / 2);
+            y_pos := y_min + (j * granularity) + (granularity / 2);
+
+            -- Create a point from x_pos and y_pos
+            point := ST_SetSRID(ST_MakePoint(x_pos, y_pos), 4326);
+
+            -- If the point is on the map calculate a score; otherwise set score to 0.
+            IF ST_Intersects(point, map_geometry) THEN
+                score := calculate_score(p_map_id, p_layer_ids, p_plant_id, date, x_pos, y_pos);
+                score := scale_score(score);  -- scale to be between 0 and 1
+                preference := score.preference;
+                relevance := score.relevance;
+            ELSE
+                preference := 0.0;
+                relevance := 0.0;
+            END IF;
+
+            x := i;
+            y := j;
+
+            RETURN NEXT;
+        END LOOP;
+    END LOOP;
+END;
+$$ LANGUAGE plpgsql;
+
+-- Scales to values between 0 and 1.
+--
+-- Preference input space: Any value.
+-- Relevance input space: >=0
+CREATE OR REPLACE FUNCTION scale_score(input SCORE)
+RETURNS SCORE AS $$
+DECLARE
+    score SCORE;
+BEGIN
+    score.preference := 1 / (1 + exp(-input.preference));       -- standard sigmoid, so that f(0)=0.5
+    score.relevance := (2 / (1 + exp(-input.relevance))) - 1;   -- modified sigmoid, so that f(0)=0
+    RETURN score;
+END;
+$$ LANGUAGE plpgsql;
+
+-- Calculate score for a certain position.
+--
+-- p_map_id       ... map id
+-- p_layer_ids[1] ... plant layer
```

Comment:
```suggestion
-- p_layer_ids[1] ... plant layer (only the first array-element is used by the function)
```

Alternatively, it could be more clear to change the array-argument to a value argument, e.g.
```SQL
p_layer_id ... id of the plant layer
```